src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+        imgs = self.src.images(start=0, num=np.inf).asnumpy()
+
+        # Obtain coefficients in polar Fourier basis for input 2D images
+        self.basis = PolarBasis2D((self.n_res, self.n_res), self.n_rad, self.n_theta)
+        self.pf = self.basis.evaluate_t(imgs)
+        self.pf = m_reshape(self.pf, (self.n_rad, self.n_theta, self.n_img))
+
+        n_theta = self.n_theta
+        if n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta = n_theta // 2
+        # The first two dimension of pf is of size n_rad x n_theta. Convert pf into an
+        # array of size (2xn_rad-1) x n_theta/2, that is, take entire ray
+        # through the origin, but take the angles only up PI. In fact, the original
+        # images are real, and thus each ray is conjugate symmetric. We therefore
+        # gain nothing by taking longer correlations (of length 2*n_rad-1 instead of n_rad),
+        # as the two halve are exactly the same. In real calculation of build_clmatrix below,
+        # We only taking shorter correlation and speed the computation by a factor of two.
+        self.pf = np.concatenate((np.flip(self.pf[1:, n_theta:], 0), self.pf[:, :n_theta]), 0)
+
+    def estimate_rotations(self):
+        """
+        Estimate orientation matrices for all 2D images
+
+        Subclasses should implement this function.
+        """
+        raise NotImplementedError('subclasses should implement this')
+
+    def save_rotations(self):
+        """
+        Save the estimated rotation matrices to new ImageSource object
+        """
+        # Generate a new ImageSource object and save estimated rotation matrices
+        # It will be simpler if it is allowed to modify the original one
+        src = ArrayImageSource(self.src.images(start=0, num=np.inf))
+        src.rots = self.rotations
+        return src
+
+    def build_clmatrix(self, n_check=None):
+        """
+        Build common-lines matrix from Fourier stack of 2D images
+
+        :param n_check: For each image/projection find its common-lines with
+            n_check images. If n_check is less than the total number of images,
+            a random subset of n_check images is used.
+        """
+
+        n_img = self.n_img
+        if n_check is None:
+            n_check = n_img
+        n_theta = self.n_theta
+        if n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta = n_theta // 2
+
+        # need to do a copy to prevent modifying self.pf for other functions
+        pf = self.pf.copy()
+
+        # Allocate local variables for return
+        # clstack represents the common lines matrix:
+        # clstack[i, j] and clstack[j ,i] contain the index
+        # of the common line of projections i and j. Namely,
+        # clstack[i,j] contains the index of the common line
+        # in the image i. clstack[j,i] contains the index of
+        # the common line in j image.
+        clstack = -np.ones((n_img, n_img))
```

Comment:
`clstack` stores the common line index which starting from 0 instead of 1 as Matlab version. -1 means there is no common line such as clstack [i,i].  Added more comments.